applications/tari_indexer/src/dan_layer_scanner.rs

```diff
@@ -50,15 +51,74 @@ impl DanLayerScanner {
     }
 
+    pub async fn get_non_fungibles(
+        &self,
+        resource_address: &ResourceAddress,
+        start_index: u64,
+        end_index: Option<u64>,
+    ) -> Result<Vec<NonFungible>, anyhow::Error> {
+        let epoch = match self.get_current_epoch().await {
+            Some(epoch) => epoch,
+            None => return Ok(vec![]),
+        };
+
+        let mut nft_substates = vec![];
+        let mut index = start_index;
+
+        loop {
+            // build the address of the nft index substate
+            let index_address = NonFungibleIndexAddress::new(*resource_address, index);
+            let index_substate_address = SubstateAddress::NonFungibleIndex(index_address);
+
+            // get the nft index substate from the network
+            // nft index substates are immutable, so they are always on version 0
+            let index_substate_result = self
+                .get_specific_substate_from_commitee(&index_substate_address, 0, epoch)
+                .await;
+            let index_substate = match index_substate_result {
+                SubstateResult::Up(substate) => substate.into_substate_value(),
+                _ => break,
+            };
+
+            // now that we have the index substate, we need the latest substate of the referenced nft
+            let nft_address = match index_substate.into_non_fungible_index() {
+                Some(idx) => idx.referenced_address().clone(),
+                // the protocol should never produce this scenario, we stop querying for more indexes if it happens
+                None => break,
+            };
+            let nft_substate_address = SubstateAddress::NonFungible(nft_address);
+            match self
+                .get_latest_substate_from_commitee(&nft_substate_address, epoch)
+                .await
+            {
+                SubstateResult::Up(substate) => {
+                    nft_substates.push(NonFungible {
+                        index,
+                        address: nft_substate_address,
+                        substate,
+                    });
+                },
+                _ => break,
+            }
+
+            if let Some(end_index) = end_index {
+                if index >= end_index {
+                    break;
+                }
+            }
+
+            index += 1;
+        }
+
+        Ok(nft_substates)
+    }
+
     pub async fn get_substate(&self, substate_address: &SubstateAddress, version: Option<u32>) -> Option<Substate> {
         info!(target: LOG_TARGET, "get_substate: {} ", substate_address);
 
-        let epoch = match self.epoch_manager.current_epoch().await {
-            Ok(epoch) => epoch,
-            Err(e) => {
-                error!(target: LOG_TARGET, "Could not retrieve the current epoch: {} ", e);
-                return None;
-            },
+        let epoch = match self.get_current_epoch().await {
+            Some(epoch) => epoch,
+            None => return None,
         };
```

Comment:
nit: can return None this way
```suggestion
        let epoch = self.get_current_epoch().await?;
```